cli: move setRawTerminal and restoreTerminal to holdHijackedConnection

coolljt0725 opened the pull request:
In this way, we can restore the Terminal as soon as possible once the hijacked
connection end. This not only fix weird output if cli enable -D, but also
remove duplicate code.

Signed-off-by: Lei Jitang leijitang@huawei.com

follow #20587 
@cpuguy83 @LK4D4 @icecrime @estesp @MHBauer 
I'll remove @MHBauer 's commit is merged.

coolljt0725 commented on api/client/hijack.go:
```diff
@@ -19,7 +33,11 @@ func (cli *DockerCli) holdHijackedConnection(tty bool, inputStream io.ReadCloser
+			restored = true
+
 			logrus.Debugf("[hijack] End of stdout")
+			logrus.Debugf("Hello world, this is a test")
```
This line is just for testing for other ones review and test this PR, will be removed once it's ready to merge

MHBauer commented:
I'll test and review this. The term handling needs some overall refactoring, methinks.

Good noticing of the duplication.

GordonTheTurtle commented:
Job: Docker-PRs-userns [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-userns/6548/console):

~~~console

---
est.go:101: DockerSuite.TestLinksNotStartedParentNotFail	0.535s
PASS: docker_cli_links_test.go:32: DockerSuite.TestLinksPingLinkedContainers	2.081s
PASS: docker_cli_links_test.go:50: DockerSuite.TestLinksPingLinkedContainersAfterRename	1.836s
PASS: docker_cli_exec_test.go:356: DockerSuite.TestLinksPingLinkedContainersOnRename	0.913s
PASS: docker_cli_links_test.go:13: DockerSuite.TestLinksPingUnlinkedContainers	0.638s

----------------------------------------------------------------------
FAIL: docker_cli_links_test.go:130: DockerSuite.TestLinksUpdateOnRestart

docker_cli_links_test.go:158:
    ip = getIP(content, "one")
docker_cli_links_test.go:144:
    c.Assert(matches, checker.NotNil, check.Commentf("Hostname %s have no matches in hosts", hostname))
... value [][]uint8 = [][]uint8(nil)
... Hostname one have no matches in hosts


----------------------------------------------------------------------
PASS: docker_cli_save_load_test.go:299: DockerSuite.TestLoadZeroSizeLayer	0.281s
PASS: docker_cli_login_test.go:11: DockerSuite.TestLoginWit
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/22093/console):

~~~console

---
unNetworkFilesBindMountROFilesystem (Test requires docker daemon to run on the same machine as CLI)
PASS: docker_cli_run_test.go:3677: DockerSuite.TestRunNetworkNotInitializedNoneMode	0.406s
PASS: docker_cli_run_test.go:493: DockerSuite.TestRunNoDupVolumes	0.104s
PASS: docker_cli_run_test.go:2218: DockerSuite.TestRunNoOutputFromPullInStdout	1.154s
PASS: docker_cli_run_test.go:3899: DockerSuite.TestRunNonExecutableCmd	0.877s

----------------------------------------------------------------------
FAIL: docker_cli_run_test.go:3910: DockerSuite.TestRunNonExistingCmd

docker_cli_run_test.go:3916:
    c.Fatalf("Run non-existing command should have errored with exit code 127, but we got exit: %d, State.ExitCode: %s", exit, stateExitCode)
... Error: Run non-existing command should have errored with exit code 127, but we got exit: 125, State.ExitCode: -1 cantstart: Cannot start container 8bfa6bf5853fd304a265b13aececde0d91339b0fa57acf691301627641e99b6c: [10] System error: invalid character 'c' looking for beginning of value



------------------------------------
---

~~~

cpuguy83 commented:
FYI, this was changed recently (adding the duplication) to deal with cases in resetting the terminal on errors during the requested action.

calavera commented:
I don't think we should do this. As @cpuguy83 mentions, the resetting is there for a good reason. We cannot have it inside the goroutine spawned to handle hijacking.

See #19513.

coolljt0725 commented:
@calavera  @cpuguy83  I think we need this. Or else if we have #20587 merged, the output of message would mess up at some case. For example(from #19506):
<pre><code>[lei@centos-188 docker]$ docker run --net=foo --ip 1.1.1.1 -it ubuntu
docker: Error response from daemon: User specified IP address is supported only when connecting to networks with user configured subnets.
                                                                                                                                         [lei@centos-188 docker]$</code></pre>

I update the fix and ti can fix the issue #19506 which fixed by #19513

icecrime commented:
There's an interesting timeout on a networking related tests on the [win2lin job](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/22444/consoleFull) (cc @mavenugo @aboch @tiborvass @jhowardmsft)

tiborvass commented:
Seems like it's `TestRunContainerNetworkModeToSelf` hanging

coolljt0725 commented:
@tiborvass  thanks for informing, fixed

icecrime commented:
@coolljt0725 Unfortunately needs a rebase already, sorry.

coolljt0725 commented:
rebased

thaJeztah commented:
ping @calavera @cpuguy83 PTAL

thaJeztah commented:
ping @calavera @cpuguy83 PTAL

LK4D4 commented on api/client/hijack.go:
```diff
@@ -8,17 +8,38 @@ import (
+			return err
+		}
+		defer func() {
+			if !restored {
```
You should use sync.Once for this stuff or race detector will be angry.

coolljt0725 replied:
@LK4D4  use sync.Once is a good idea :+1:

LK4D4 commented on api/client/run.go:
```diff
@@ -228,6 +224,17 @@ func (cli *DockerCli) CmdRun(args ...string) error {
+		// If we have holdHijackedConnection, we should notify
+		// holdHijackedConnection we are going to exit and wait
+		// to avoid the terminal are not restored.
+		if config.AttachStdin || config.AttachStdout || config.AttachStderr {
```
let's create variable for this

LK4D4 commented on api/client/run.go:
```diff
@@ -228,6 +224,17 @@ func (cli *DockerCli) CmdRun(args ...string) error {
+			select {
+			case <-exitCh:
+			default:
+				close(exitCh)
```
It's only place where this channel can be closed, so let's use just `close(exitCh)`

coolljt0725 replied:
@LK4D4  Just in case we may close channel somewhere in future and thus will leave a bug here, so it's much more safer to check if the channel has been closed before closing. but I'm ok with removeingthis check.

LK4D4 commented on api/client/run.go:
```diff
@@ -228,6 +224,17 @@ func (cli *DockerCli) CmdRun(args ...string) error {
+			default:
+				close(exitCh)
+			}
+			<-errCh
```
Doesn't it makes sense at least log error here if it's not nil?

coolljt0725 replied:
@LK4D4  we already have a `Debugf` in `holdHijackedConnection` if `err` is not nil, so I think there is no need here.

LK4D4 commented on api/client/start.go:
```diff
@@ -86,7 +86,10 @@ func (cli *DockerCli) CmdStart(args ...string) error {
-		var in io.ReadCloser
+		var (
+			in     io.ReadCloser
+			exitCh chan struct{}
```
unnecessary declaration

LK4D4 commented on api/client/run.go:
```diff
@@ -158,6 +158,7 @@ func (cli *DockerCli) CmdRun(args ...string) error {
 	var (
 		waitDisplayID chan struct{}
 		errCh         chan error
+		exitCh        chan struct{}
```
unnecessary declaration

coolljt0725 replied:
we have to declare `exitCh` out of the `if !config.AttachStdout && !config.AttachStderr` block, since we use it out of this block

LK4D4 commented on api/client/hijack.go:
```diff
@@ -8,17 +8,38 @@ import (
 
-func (cli *DockerCli) holdHijackedConnection(tty bool, inputStream io.ReadCloser, outputStream, errorStream io.Writer, resp types.HijackedResponse) error {
-	var err error
+func (cli *DockerCli) holdHijackedConnection(tty bool, inputStream io.ReadCloser, outputStream, errorStream io.Writer, resp types.HijackedResponse, exitCh chan struct{}) error {
```
we probably should use net/context for this. WDYT @calavera ?

coolljt0725 commented:
@LK4D4  updated

LK4D4 commented:
@coolljt0725 thanks! I still think that it might be a suitable job for net/context. Also, we're adding its support to all client methods in #21219

coolljt0725 commented:
@LK4D4  that's cool :-)

coolljt0725 commented:
@LK4D4   #21219 is merged, but I think this is still need, or else the output would be a bit of weird if we failed to run a container
<pre><code>[lei@centos-188 docker]$ docker run --net=foo --ip 1.1.1.1 -it ubuntu
docker: Error response from daemon: User specified IP address is supported only when connecting to networks with user configured subnets.
                                                                                                                                         [lei@centos-188 docker]$</code></pre>
I rebased `holdHijackedConnection` with `context`. 
@cpuguy83  @calavera

LK4D4 commented:
LGTM

cpuguy83 commented on api/client/hijack.go:
```diff
@@ -2,23 +2,48 @@ package client
+func (cli *DockerCli) holdHijackedConnection(ctx context.Context, tty bool, inputStream io.ReadCloser, outputStream, errorStream io.Writer, resp types.HijackedResponse) error {
+	var (
+		err  error
+		once sync.Once
```
It'd be nice to have a better var name here, like `restoreOnce`

cpuguy83 commented:
One nit in the code, otherwise LGTM

coolljt0725 commented:
@cpuguy83  done

coolljt0725 committed:
cli: move setRawTerminal and restoreTerminal to holdHijackedConnection

In this way, we can restore the Terminal as soon as possible once the hijacked
connection end. This not only fix weird output if cli enable -D, but also
remove duplicate code.

Signed-off-by: Lei Jitang <leijitang@huawei.com>

LK4D4 commented:
I think it's good to merge

coolljt0725 commented:
ping @thaJeztah  this should go in 1.11.0 ?

thaJeztah commented:
I asked @LK4D4 and he wasn't sure, but perhaps I interpreted incorrectly

/cc @tiborvass
